src/cockpit/remote.py

```
@@ -15,21 +15,185 @@
+        try:
+            logger.debug('checking ~/.ssh/known_hosts')
+            with open(os.path.expanduser('~/.ssh/known_hosts')) as file:
+                logger.debug('  file is open')
+                if key in file.read():
+                    logger.debug('  and contains the key!')
+                    return False  # this looks like it's going to succeed on its own
+                logger.debug('  but does not contain the key :(')
+        except OSError as exc:
+            logger.debug('  err: %s', exc)
+
+        raise PeerError('unknown-hostkey', error='unknown-hostkey', auth_method_results={},
+                        host_key=f'{host} {algorithm} {key}', host_fingerprint=fingerprint)
+
+    async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[str]:
+        logger.debug('Got askpass(%s): %s', hint, prompt)
+
+        match = PasswordResponder.PASSPHRASE_RE.fullmatch(prompt)
+        if match is not None:
+            path = match.group(1)
+            logger.debug("This is a passphrase request for %s, but we don't do those.  Abort.", path)
+            self.error_message = f'locked identity: {path}'
+            return None
+
+        assert self.password is not None
+        assert self.attempts == 0
+        self.attempts += 1
+        return self.password
+
+
+class SshPeer(Peer):
+    session: Optional[ferny.Session] = None
+    host: str
+    user: Optional[str]
+    password: Optional[str]
+    private: bool
+
+    async def do_connect_transport(self) -> asyncio.Transport:
+        assert self.session is not None
+        logger.debug('Starting ssh session user=%s, host=%s, private=%s', self.user, self.host, self.private)
+
+        basename, colon, portstr = self.host.rpartition(':')
+        if colon and portstr.isdigit():
+            host = basename
+            port = int(portstr)
+        else:
+            host = self.host
+            port = None
```

This silently ignores cases like `self.host == 'foo:bar'`. Then `colon == ':` but the isdigit() test fails. This case should throw an error, instead of silently ignoring the `:bar` part.

---

I don't agree that the port will be dropped in this case.  It will hit the `else:` branch and set `host = self.host` (ie: the full original name) and `port = None`.

---

Oh, ok - that feels rather strange and unpredictable, but hmm..